### PR TITLE
Global clang-format

### PR DESCRIPTION
**Pull request by rwgk**
**EXPERIMENTAL** integration of clang-format hook into pre-commit.
**NOT TO BE MERGED** until after the 2.7 release.

At the moment this DRAFT PR is is to see what's involved and what the result looks like.

Using the same docker container as used for clang-tidy (silkeh/clang:12).

### CODE REVIEW

**Comment by Skylion007** [comment]
@rwgk Could you hold off on this for a sec, there are some readability clang-tidy checks that will also enforce LLVM style more strictly and it would be better to be one massive formatting diff than two.

**Comment by rwgk** [comment]
> @rwgk Could you hold off on this for a sec, there are some readability clang-tidy checks that will also enforce LLVM style more strictly and it would be better to be one massive formatting diff than two.

Absolutely: see the updated PR description.

**Comment by rwgk** [comment]
Note to self: reverting the clang-format changes in these files did NOT resolve the MSVC 2015 failures:
```
diff --combined include/pybind11/chrono.h
diff --combined include/pybind11/pybind11.h
diff --combined tests/test_chrono.cpp
```

File: include/pybind11/pytypes.h [hunk 1]
```diff
@@ -47,54 +47,65 @@ using tuple_accessor = accessor<accessor_policies::tuple_item>;
...
         the returned ``detail::item_accessor`` instance to a `handle` or `object`
         subclass causes a corresponding call to ``__getitem__``. Assigning a `handle`
         or `object` subclass causes a call to ``__setitem__``.
     \endrst */
+    // clang-format on
     item_accessor operator[](handle key) const;
     /// See above (the only difference is that they key is provided as a string literal)
     item_accessor operator[](const char *key) const;
 
+    // clang-format off
```
Comment (Skylion007) [inline include/pybind11/pytypes.h:88]: Why is this necessary?

Comment (rwgk) [reply include/pybind11/pytypes.h:88]: To protect the `\rst` to `\endrst` block enclosed by the `// clang-format off`, `on`.

I added the clang-format meta-comments systematically rather than trying to figure out case by case if they are needed or not. Possibly some could be omitted, but I believe for blocks that are deliberately formatted as rst it's better to systematically exclude them from clang-format.

Note that the sphinx error messages are completely useless if something goes wrong. I spent quite a bit of time figuring out why the Documentation GitHub action was failing. Now I know, but letting clang-format loose on rst is likely to turn into time-consuming surprises for others.

File: include/pybind11/eval.h [hunk 2]
```diff
@@ -109,41 +115,46 @@ object eval_file(str fname, object global = globals(), object local = object())
...
+            start = Py_file_input;
+            break;
+        default:
+            pybind11_fail("invalid evaluation mode");
     }
 
     int closeFile = 1;
     std::string fname_str = (std::string) fname;
-#if PY_VERSION_HEX >= 0x03040000
+#    if PY_VERSION_HEX >= 0x03040000
```
[the remark this answers is not in the document]
Comment (rwgk) [reply include/pybind11/eval.h:133]: Good question.
I was thinking of formatting globally in different ways (multiple PRs), then we can look at the results holistically.

BTW: I'm still hung up on one last (I hope) thing, MSVC 2015 failures after clang-format.

[the remark this answers is not in the document]
Comment (rwgk) [reply include/pybind11/eval.h:133]: BTW: I fixed the MSVC 2015 issue.

To your style question, there are only 3 choices: https://clang.llvm.org/docs/ClangFormatStyleOptions.html
`PPDIS_None`, `PPDIS_AfterHash`, `PPDIS_BeforeHash`

What we have (AfterHash) seems like the most OK-ish option to me.
But really, any of these is better than having to mess around with the whitespaces manually.

**Comment by rwgk** [comment]
This PR is completely stale, even as a demo. The only slightly valuable commit is in the diff below. Closing & deleting branch.

```diff
diff --git a/.pre-commit-config.yaml b/.pre-commit-config.yaml
index 133eb6df4e..6af90465ae 100644
--- a/.pre-commit-config.yaml
+++ b/.pre-commit-config.yaml
@@ -12,6 +12,10 @@
 #
 # See https://github.com/pre-commit/pre-commit
 
+ci:
+  skip:
+  - docker-clang-format
+
 repos:
 # Standard hooks
 - repo: https://github.com/pre-commit/pre-commit-hooks
@@ -108,6 +112,19 @@ repos:
     entry: PyBind|Numpy|Cmake|CCache
     exclude: .pre-commit-config.yaml
 
+- repo: local
+  hooks:
+  - id: docker-clang-format
+    name: Docker Clang Format
+    language: docker_image
+    types:
+    - c++
+    entry: silkeh/clang:12
+    args:
+    - clang-format
+    - -style=file
+    - -i
+
 - repo: local
   hooks:
   - id: check-style
```
